### ISSUE TITLE
fluid quit ('x' button) should check if Shell -> Execute still running and warn instead of hang

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#850]
**Describe the bug**
If one uses Shell -> Execute to build and run an app, leaves the built app running and hits 'X' to close fluid, fluid "hangs" on the waitpid() for the process to finish, triggering the window manager's "Force Quit" dialog, e.g.

![image](https://github.com/fltk/fltk/assets/6484779/5aa17ab7-3838-4f7a-81fc-37a9fca99fe8)

This could be avoided by having fluid check if there's an "execute command" still running, and post a warning dialog if so indicating the Shell -> Execute command is still running (I think the old fluid used to do this?), with the options "Cancel" (abort quit), "Quit (Try again)", and perhaps a third option to "Force Quit" which kills the executing command.

The workaround is closing the built app, which completes the Shell -> Executing command,
**but it's not obvious that's the problem causing the hang.**

Here's the calling stack for the hang showing it waiting for the waitpid():
```
(gdb) where
#0  0x00007f0304aea42f in __GI___wait4 (pid=108343, stat_loc=stat_loc@entry=0x7ffd630970ac, options=options@entry=0, usage=usage@entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:30
#1  0x00007f0304aea3ab in __GI___waitpid (pid=<optimized out>, stat_loc=stat_loc@entry=0x7ffd630970ac, options=options@entry=0) at ./posix/waitpid.c:38
#2  0x00007f0304a8090a in _IO_new_proc_close (fp=0x12785f0) at ./libio/iopopen.c:285
#3  0x00007f0304a8bf00 in _IO_new_file_close_it (fp=fp@entry=0x12785f0) at ./libio/libioP.h:947
#4  0x00007f0304a7ed8f in _IO_new_fclose (fp=0x12785f0) at ./libio/iofclose.c:53
#5  0x000000000047c105 in Fl_Process::close() ()
#6  0x000000000047c0a4 in Fl_Process::~Fl_Process() ()
#7  0x00007f0304a45495 in __run_exit_handlers (status=0, listp=0x7f0304c19838 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at ./stdlib/exit.c:113
#8  0x00007f0304a45610 in __GI_exit (status=<optimized out>) at ./stdlib/exit.c:143
#9  0x0000000000412593 in exit_cb(Fl_Widget*, void*) ()
#10 0x000000000051ffce in Fl_Widget::do_callback(Fl_Widget*, void*, Fl_Callback_Reason) ()
#11 0x00000000004172c4 in Fl_Widget::do_callback(Fl_Callback_Reason) ()
#12 0x0000000000492ccc in Fl::handle_(int, Fl_Window*) ()
#13 0x0000000000492c0d in Fl::handle(int, Fl_Window*) ()
#14 0x000000000057a1a7 in delayed_close(Fl_Window*) ()
#15 0x0000000000491b4b in Fl::run_checks() ()
#16 0x00000000004e985e in Fl_System_Driver::wait(double) ()
#17 0x0000000000580463 in Fl_Unix_System_Driver::wait(double) ()
#18 0x0000000000491d4d in Fl::wait(double) ()
#19 0x0000000000491db4 in Fl::wait() ()
#20 0x0000000000416cc6 in main ()
(gdb) 
```

**Comment by MatthiasWM** [comment]
I check what I can do. I had Fl_Process hang for me too before. 

We talked about using a better class to run system applications. You said that you have a class that includes killing processes and their spawned processes. Would it make sense to add this to Fl_Process? Do you have code I could use?

**Comment by erco77** [comment]
There is code in fluid for when I added the external editor stuff that keeps track of the PID, but I would think you can just check if a process is still running using whatever's in Fl_Process.. i.e. checking if the _fpt member is NULL?

**Comment by erco77** [comment]
Attaching a patch suggestion -- this kept tripping me up, lol
[fluid-patch-v1.txt](https://github.com/fltk/fltk/files/13501781/fluid-patch-v1.txt)

**Comment by MatthiasWM** [comment]
@erco77 Thanks, yes, until we are actually able to kill the process, this is the best thing to do.

**Comment by MatthiasWM** [comment]
@erco77 I am applying the patch now and extend it a little, so no need for you to do it

**Comment by MatthiasWM** [comment]
Fixed in 2025481f8c19fd73a611e517a0c91c15d7ed390a

**Comment by erco77** [comment]
*yawn* Ah, ok.. great!
